exporter/kafkaexporter/kafka_exporter.go

```diff
@@ -265,23 +265,34 @@ func (e *kafkaLogsMessenger) marshalData(ld plog.Logs) ([]marshaler.Message, err
 }
 
 func (e *kafkaLogsMessenger) getTopic(ctx context.Context, ld plog.Logs) string {
-	return getTopic(ctx, e.config.Logs, e.config.TopicFromAttribute, ld.ResourceLogs())
+	return getTopic[plog.ResourceLogs](ctx, e.config.Logs, e.config.TopicFromAttribute, ld.ResourceLogs())
 }
 
 func (e *kafkaLogsMessenger) partitionData(ld plog.Logs) iter.Seq2[[]byte, plog.Logs] {
 	return func(yield func([]byte, plog.Logs) bool) {
-		if !e.config.PartitionLogsByResourceAttributes {
-			yield(nil, ld)
+		if e.config.PartitionLogsByResourceAttributes {
+			for _, resourceLogs := range ld.ResourceLogs().All() {
+				hash := pdatautil.MapHash(resourceLogs.Resource().Attributes())
+				newLogs := plog.NewLogs()
+				resourceLogs.CopyTo(newLogs.ResourceLogs().AppendEmpty())
+				if !yield(hash[:], newLogs) {
+					return
+				}
+			}
 			return
 		}
-		for _, resourceLogs := range ld.ResourceLogs().All() {
-			hash := pdatautil.MapHash(resourceLogs.Resource().Attributes())
-			newLogs := plog.NewLogs()
-			resourceLogs.CopyTo(newLogs.ResourceLogs().AppendEmpty())
-			if !yield(hash[:], newLogs) {
-				return
+		if e.config.PartitionLogsByTraceID {
+			for _, l := range batchpersignal.SplitLogs(ld) {
+				key := []byte(traceutil.TraceIDToHexOrEmptyString(
+					l.ResourceLogs().At(0).ScopeLogs().At(0).LogRecords().At(0).TraceID(),
+				))
+				if !yield(key, l) {
+					return
+				}
```

Comment:
BTW I think this will partition all log records without a trace ID to the same partition. To fall back to the default partitioning, you should check if it has a trace ID and pass `nil` in as the message key in case it does not.